File: java/src/test/java/br/com/vps/consulting/b2b/management/partner/application/usecase/list/DefaultListPartnersUseCaseTest.java
package br.com.vps.consulting.b2b.management.partner.application.usecase.list;

import br.com.vps.consulting.b2b.management.partner.domain.Partner;
import br.com.vps.consulting.b2b.management.partner.domain.PartnerId;
import br.com.vps.consulting.b2b.management.partner.domain.PartnerRepository;
import br.com.vps.consulting.b2b.management.shared.core.page.PageCustom;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class DefaultListPartnersUseCaseTest {

    @Mock private PartnerRepository partnerRepository;
    @InjectMocks private DefaultListPartnersUseCase useCase;

    @Test
    @DisplayName("Should return mapped PageCustom<PartnerListOutput> from repository page")
    void shouldReturnMappedPage() {
        final var partner = aPartner();
        final var page = PageCustom.<Partner>builder()
                .pageNumber(0).pageSize(10).numberOfElements(1)
                .totalPages(1).totalElements(1L).items(List.of(partner))
                .build();
        when(partnerRepository.findAll(10, 0)).thenReturn(page);

        final var result = useCase.execute(new ListPartnersInput(10, 0));

        assertThat(result.items()).hasSize(1);
        assertThat(result.totalElements()).isEqualTo(1L);
        assertThat(result.pageNumber()).isEqualTo(0);
        assertThat(result.items().get(0).id()).isEqualTo(partner.getId().value());
        assertThat(result.items().get(0).name()).isEqualTo(partner.getName());
        assertThat(result.items().get(0).document()).isEqualTo(partner.getDocument());
    }

    @Test
    @DisplayName("Should forward all input parameters to repository")
    void shouldPassInputParamsToRepository() {
        when(partnerRepository.findAll(5, 2)).thenReturn(emptyPage());

        useCase.execute(new ListPartnersInput(5, 2));

        verify(partnerRepository).findAll(5, 2);
    }

    @Test
    @DisplayName("Should return empty page when no partners exist")
    void shouldReturnEmptyPageWhenNoPartners() {
        when(partnerRepository.findAll(20, 0)).thenReturn(emptyPage());

        final var result = useCase.execute(new ListPartnersInput(20, 0));

        assertThat(result.items()).isEmpty();
        assertThat(result.totalElements()).isZero();
    }

    private static Partner aPartner() {
        return Partner.builder()
                .id(PartnerId.generate())
                .name("Acme Corp")
                .document("12345678000100")
                .creditLimit(new BigDecimal("5000.00"))
                .createdAt(Instant.now())
                .build();
    }

    private static PageCustom<Partner> emptyPage() {
        return PageCustom.<Partner>builder()
                .pageNumber(0).pageSize(20).numberOfElements(0)
                .totalPages(0).totalElements(0L).items(List.of())
                .build();
    }
}

File: kotlin/src/test/kotlin/br/com/vps/consulting/b2b/management/partner/application/usecase/list/DefaultListPartnersUseCaseTest.kt
package br.com.vps.consulting.b2b.management.partner.application.usecase.list

import br.com.vps.consulting.b2b.management.partner.domain.Partner
import br.com.vps.consulting.b2b.management.partner.domain.PartnerRepository
import br.com.vps.consulting.b2b.management.shared.core.page.PageCustom
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever

@ExtendWith(MockitoExtension::class)
class DefaultListPartnersUseCaseTest {

    @Mock
    lateinit var partnerRepository: PartnerRepository

    private lateinit var useCase: DefaultListPartnersUseCase

    @BeforeEach
    fun setUp() {
        useCase = DefaultListPartnersUseCase(partnerRepository)
    }

    @Test
    fun `should return mapped PageCustom of ListPartnersOutput from repository page`() {
        val partner = Partner.create(name = "Acme Corp", document = "12345678000100")
        val page = PageCustom(pageNumber = 0, pageSize = 20, totalPages = 1, totalElements = 1L, items = listOf(partner))
        whenever(partnerRepository.findAll("12345678000100", 20, 0)).thenReturn(page)

        val result = useCase.execute(ListPartnersInput(document = "12345678000100", pageSize = 20, pageNumber = 0))

        assertThat(result.items).hasSize(1)
        assertThat(result.items[0].id).isEqualTo(partner.id.value)
        assertThat(result.items[0].name).isEqualTo("Acme Corp")
        assertThat(result.totalElements).isEqualTo(1L)
    }

    @Test
    fun `should forward document pageSize and pageNumber from input to repository findAll`() {
        whenever(partnerRepository.findAll(null, 10, 2))
            .thenReturn(PageCustom(pageNumber = 2, pageSize = 10, totalPages = 0, totalElements = 0L, items = emptyList()))

        useCase.execute(ListPartnersInput(document = null, pageSize = 10, pageNumber = 2))

        verify(partnerRepository).findAll(null, 10, 2)
    }

    @Test
    fun `should return empty page when no partners exist`() {
        whenever(partnerRepository.findAll(null, 20, 0))
            .thenReturn(PageCustom(pageNumber = 0, pageSize = 20, totalPages = 0, totalElements = 0L, items = emptyList()))

        val result = useCase.execute(ListPartnersInput(document = null, pageSize = 20, pageNumber = 0))

        assertThat(result.items).isEmpty()
        assertThat(result.totalElements).isZero()
    }

}
